Floor Selector feature added

cristian-situm opened the pull request.

cristian-situm committed:
Floor Selector feature added

R-lagoman commented on app/src/main/java/es/situm/gettingstarted/drawbuilding/FloorAdapter.java:
```diff
@@ -0,0 +1,97 @@
+import es.situm.sdk.model.cartography.Floor;
+
+/**
+ * Created by adriannieto on 5/10/17.
```
Este comentario non aplica.

R-lagoman commented on app/src/main/java/es/situm/gettingstarted/drawbuilding/FloorAdapter.java:
```diff
@@ -0,0 +1,97 @@
+/**
+ * Created by adriannieto on 5/10/17.
+ */
+@VisibleForTesting(otherwise = VisibleForTesting.PACKAGE_PRIVATE)
```
Hai moitas anotacións `@VisibleForTesting` pero non vexo que se estea a utilizar esta clase en ningún test (hai tests?). Entendo que sobran.

R-lagoman commented on app/src/main/java/es/situm/gettingstarted/drawbuilding/DrawBuildingActivity.java:
```diff
@@ -65,6 +81,15 @@ public void onError(Error error) {
+        selectorMarkTop = findViewById(R.id.situm_floor_selector_mark_top);
+        selectorMarkBottom = findViewById(R.id.situm_floor_selector_mark_bottom);
+
+        floorAdapter = new FloorAdapter(Collections.<Floor>emptyList(), floor -> {/* do nothing */});
```
Por que se instancia aquí `FloorAdapter` utilizando un listener baleiro?
Volves a instancialo na liña 131, esta vez pasando o listener. 
Si o que che preocupa é que o adapter sexa `null`, podes instanciar pasando `emptyList` e un listener válido e posteriormente utilizar `setFloors`, que polo que vexo está implementado no adapter.
PD: o listener non ten por que ser unha clase anónima, podes facer que `DrawBuildingActivity` implemente a interface, atributo privado, etc.
PD 2: fai falta parametrizar `Collections.<Floor>emptyList()`? Diría que o tipo xa pode ser inferido.

cristian-situm replied:
Corrección: pasei a instancialo 1 soa vez nun novo método "prepareSelector()"

R-lagoman commented on app/src/main/java/es/situm/gettingstarted/drawbuilding/FloorAdapter.java:
```diff
@@ -0,0 +1,97 @@
+
+    public void select(Floor floor) {
+        selected = floor;
+        notifyDataSetChanged();
```
Probablemente neste caso non importe, pero en conxuntos de datos maiores sería mellor ser máis específico á hora de notificar que cambiou o dataset do adapter.

cristian-situm replied:
Estuve revisando los notify de RecyclerView y no encontré otro con el que sustituírlo. El que me pareció el más adecuado fue el de **notifyItemChanged()** que cambia el background del item seleccionado pero no el de los demás items, entonces acaban todos seleccionados

![Screenshot from 2022-04-12 12-08-19](https://user-images.githubusercontent.com/103040497/162936431-c5309ce7-17ba-4b9a-8d2e-26732e3e3f09.png)
.

R-lagoman commented on app/src/main/java/es/situm/gettingstarted/drawbuilding/DrawBuildingActivity.java:
```diff
@@ -83,23 +108,71 @@ void drawBuilding(Bitmap bitmap) {
         SitumSdk.communicationManager().fetchFloorsFromBuilding(building.getIdentifier(), new Handler<Collection<Floor>>() {
             @Override
             public void onSuccess(Collection<Floor> floorsCollection) {
                 List<Floor> floors = new ArrayList<>(floorsCollection);
```
Non entendo que fan aquí varias cousas:

1. Nesta liña creamos un novo `ArrayList` a partir da `Collection` que xa tiñamos.
2. Na 118 instanciamos un `LinkedHashMap` que ademais é un atributo da clase.
3. A continuación poboamos ese map cos datos que xa tiñamos na `Collection` e tamén na `List`, pero metemos os identificadores de planta como claves.
4. Os identificadores de planta parece que non se usan nunca, para que os queremos?
5. Na 131 volvemos a crear un novo `ArrayList` a partir dos `values` do map?¿?¿?
Por que todo esto? Non podes instanciar o `FloorAdapter` partindo da `Collection` inicial?

cristian-situm replied:
Aquí non parei 2 veces a mirar o código a fondo. Cambiei toda esta parafernalia por dúas listas:

- Un List<> para poder iterar con índices a non un iterator.
- Un LinkedList<> para meter os anteriores datos en orden inverso e por orden de chegada.

R-lagoman commented on app/src/main/java/es/situm/gettingstarted/drawbuilding/DrawBuildingActivity.java:
```diff
@@ -2,6 +2,13 @@
 
```
Recentemente engadiuse un ficheiro `CHANGELOG`, habería que escribir notas no historial de cambios.

R-lagoman commented on app/src/main/java/es/situm/gettingstarted/drawbuilding/DrawBuildingActivity.java:
```diff
@@ -83,23 +108,71 @@ void drawBuilding(Bitmap bitmap) {
+                        @Override
+                        public void onSuccess(Bitmap bitmap) {
+                            callback.onSuccess(bitmap);
+                            floorAdapter.select(floor);
```
Entendo que o adapter é notificado de que cambiou a planta só cando xa se obtivo o bitmap. Eso quere dicir que cando a rede é lenta tocarás no botón para cambiar de planta e non sucederá nada ata 10 ou 15 segundos máis tarde?
Si é así, aquí vexo que hai 3 cousas que deberían terse en conta por separado:
1. O que sucede localmente que podemos controlar sin agardar pola imaxe.
2. O que sucede mentres se carga a imaxe (neste caso non hai loadings nin nada, pero puidera darse o caso e incluso non o descarto).
3. O que sucede posteriormente, cando por fin temos a imaxe.
Como o ves?

R-lagoman commented on app/src/main/java/es/situm/gettingstarted/drawbuilding/DrawBuildingActivity.java:
```diff
@@ -49,10 +65,10 @@ protected void onCreate(Bundle savedInstanceState) {
+        fetchFloorsImages(building, new Callback() {
             @Override
             public void onSuccess(Bitmap floorImage) {
                 drawBuilding(floorImage);
```
Diría que `drawBuilding` xa non é "draw building", non? Máis ben sería un "draw floor" ou algo así?

R-lagoman commented on app/src/main/java/es/situm/gettingstarted/drawbuilding/DrawBuildingActivity.java:
```diff
@@ -83,23 +108,71 @@ void drawBuilding(Bitmap bitmap) {
+                Floor defaultFloor = floors.get(0);
+
+                // The next method selects by default the first item of the List<>
+                SitumSdk.communicationManager().fetchMapFromFloor(defaultFloor, new Handler<Bitmap>() {
```
Esta segunda chamada ao método `fetchMapFromFloor` deberías poder evitala.
Estás adaptando o que había antes, que pescaba unha planta do edificio e a pintaba. Eso é natural, pero neste caso o refactor creo que pide un pouquiño máis.
Para min o natural sería:
1. Partir dun conxunto baleiro de datos e inicializar todo (adapters, listeners, o que sexa): o código debe saber comportarse aínda que haxa 0 plantas.
2. Pedir as plantas do edificio. Unha vez as tes, actualizas o teu _underlying dataset_.
3. Cando xa tes datos, seleccionas unha planta por defecto, utilizando os mecanismos que xa tes instanciados, non unha nova chamada a `fetchMapFloor`.

R-lagoman commented on app/src/main/java/es/situm/gettingstarted/drawbuilding/DrawBuildingActivity.java:
```diff
@@ -83,23 +108,71 @@ void drawBuilding(Bitmap bitmap) {
+                SitumSdk.communicationManager().fetchMapFromFloor(defaultFloor, new Handler<Bitmap>() {
                     @Override
                     public void onSuccess(Bitmap bitmap) {
                         callback.onSuccess(bitmap);
```
Igual que comentei arriba: antes pode ser que pintara algo, pero revisa si esta `callback` pinta algo agora.

cristian-situm committed:
1. Reorganization of the code
2. Issue in drawFloor() fixed by removing the current floor image for the new one.
3. CHANGELOG.md updated

cristian-situm committed:
1. Reorganization of the code
2. Issue in drawFloor() fixed by removing the current floor image for the new one.
3. CHANGELOG.md updated
